src/server/editorServices.ts

```diff
@@ -3673,6 +3686,96 @@ namespace ts.server {
+                    return true;
+                }
+            });
+
+            return result;
+        }
+
+        /*@internal*/
+        private watchPackageJsonFile(path: Path, watchOptions: WatchOptions | undefined) {
+            const watchers = this.packageJsonFilesMap || (this.packageJsonFilesMap = createMap());
+            if (!watchers.has(path)) {
+                this.invalidateProjectAutoImports(path);
+                watchers.set(path, this.watchFactory.watchFile(
+                    this.host,
+                    path,
+                    (fileName, eventKind) => {
+                        const path = this.toPath(fileName);
+                        switch (eventKind) {
+                            case FileWatcherEventKind.Created:
+                                return Debug.fail();
+                            case FileWatcherEventKind.Changed:
+                                this.packageJsonCache.addOrUpdate(path);
+                                this.invalidateProjectAutoImports(path);
+                                break;
+                            case FileWatcherEventKind.Deleted:
+                                this.packageJsonCache.delete(path);
+                                this.invalidateProjectAutoImports(path);
+                                watchers.get(path)!.close();
+                                watchers.delete(path);
+                        }
+                    },
+                    PollingInterval.Low,
+                    watchOptions,
+                    WatchType.PackageJsonFile,
+                ));
+            }
+        }
+
+        /*@internal*/
+        private onAddPackageJson(path: Path, watchOptions: WatchOptions | undefined) {
+            this.packageJsonCache.addOrUpdate(path);
+            this.watchPackageJsonFile(path, watchOptions);
+        }
+
+        /*@internal*/
+        includePackageJsonAutoImports(): PackageJsonAutoImportPreference {
+            switch (this.hostConfiguration.preferences.includePackageJsonAutoImports) {
+                case "none": return PackageJsonAutoImportPreference.None;
+                case "all": return PackageJsonAutoImportPreference.All;
+                default: return PackageJsonAutoImportPreference.ExcludeDevDependencies;
+            }
+        }
+
+        /*@internal*/
+        private invalidateProjectAutoImports(packageJsonPath: Path | undefined) {
+            if (this.includePackageJsonAutoImports()) {
+                this.configuredProjects.forEach(invalidate);
+                this.inferredProjects.forEach(invalidate);
+                this.externalProjects.forEach(invalidate);
+            }
+            function invalidate(project: Project) {
+                if (!packageJsonPath || project.packageJsonsForAutoImport?.has(packageJsonPath)) {
+                    project.getImportSuggestionsCache().clear();
+                    project.markAutoImportProviderAsDirty();

```

Comment:
Do  you think getImportSuggestionsCache should clear from AutoImportProviderProject::markAsDirty instead? 
Also do you think it should clear if there is program change in AutoImportProvider instead.

Reply:
No, because `AutoImportProvider` only includes files from `node_modules` that aren’t in the main program, whereas the completions cache contains all auto imports, including from local files.

Reply:
So it main project contains auto import suggestions from auto import provider too? If yes when we souvenir file in auto import provider changes or say resolved module changes, shouldn’t it clear the cache of main project ? If not what do we need to clear it from invalidate when host config changes for the import preferences

Reply:
If the structure of the main program changes, we already clear the cache. The only way the structure of the auto import program can change is if the structure of the main program changes (which is handled), a change occurs in node_modules or a package.json (which is handled right here), or if project/user settings change—there could possibly be a bug there, but it would be a year-old bug that no one has ever reported.

Reply:
WatchFailedLookupLocations, onSourceFileChanged, WatchMissingFile, watchTypeRootsDirectory, dont call this though.. which is why i think probably you want to clear importSuggestionCache whenever there is program update in the AutoImportProvider.. Again if it depends on only fileNames in the program then you can use filesAddedOrRemoved in updateGraphWorker otherwise hasNewProgram

Reply:
Ah, I see what you mean. Ok, I think that makes sense then.